homeassistant/components/trafikverket_train/sensor.py

```diff
@@ -117,98 +128,99 @@ def next_departuredate(departure):
+        try:
+            if self._time:
                 self._state = await self._train_api.async_get_train_stop(
                     self._from_station, self._to_station, when
                 )
-            except ValueError as output_error:
-                _LOGGER.error(
-                    "Departure %s encountered a problem: %s", when, output_error
+            else:
+
+                self._state = await self._train_api.async_get_next_train_stop(
+                    self._from_station, self._to_station, when
                 )
-        else:
-            when = datetime.now()
-            self._state = await self._train_api.async_get_next_train_stop(
-                self._from_station, self._to_station, when
-            )
-        self._departure_state = self._state.get_state().name
-        self._delay_in_minutes = self._state.get_delay_time()
+        except ValueError as output_error:
+            _LOGGER.error("Departure %s encountered a problem: %s", when, output_error)
 
-    @property
-    def extra_state_attributes(self):
-        """Return the state attributes."""
         if not self._state:
-            return None
-        attributes = {
-            ATTR_DEPARTURE_STATE: self._departure_state,
-            ATTR_CANCELED: self._state.canceled,
-            ATTR_DELAY_TIME: None,
-            ATTR_PLANNED_TIME: None,
-            ATTR_ESTIMATED_TIME: None,
-            ATTR_ACTUAL_TIME: None,
-            ATTR_OTHER_INFORMATION: None,
-            ATTR_DEVIATIONS: None,
-        }
-        if self._state.other_information:
-            attributes[ATTR_OTHER_INFORMATION] = ", ".join(
-                self._state.other_information
-            )
-        if self._state.deviations:
-            attributes[ATTR_DEVIATIONS] = ", ".join(self._state.deviations)
-        if self._delay_in_minutes:
-            attributes[ATTR_DELAY_TIME] = self._delay_in_minutes.total_seconds() / 60
-        if self._state.advertised_time_at_location:
-            attributes[ATTR_PLANNED_TIME] = as_utc(
-                self._state.advertised_time_at_location.astimezone(self._timezone)
-            ).isoformat()
+            self._attr_available = False
+            self._attr_native_value = None
+            self._attr_extra_state_attributes = {}
+            return
+
+        self._attr_available = True
+
+        self._attr_native_value = self._state.advertised_time_at_location.astimezone(
+            TIMEZONE
+        )
+        if self._state.time_at_location:
+            self._attr_native_value = self._state.time_at_location.astimezone(TIMEZONE)
         if self._state.estimated_time_at_location:
-            attributes[ATTR_ESTIMATED_TIME] = as_utc(
-                self._state.estimated_time_at_location.astimezone(self._timezone)
+            self._attr_native_value = self._state.estimated_time_at_location.astimezone(
+                TIMEZONE
+            )
+
```

Comment:
Maybe split here and create a new method `_update_attributes` to make it less heavy to the eyes.

Reply:
There is probably some code prettifying possible on the attributes to reduce multiline stuff.
I'll have a think

Reply:
I have thought some more about it, and I think the original way to do it was cleaner: initialise with default values, and then use `if ` checks to populate the extra attributes.

```python
    self._attr_extra_state_attributes = {
            ATTR_DEPARTURE_STATE: self._state.get_state().name,
            ATTR_CANCELED: self._state.canceled,
            ATTR_DELAY_TIME: None,
            ATTR_PLANNED_TIME: None,
            ATTR_ESTIMATED_TIME: None,
            ATTR_ACTUAL_TIME: None,
            ATTR_OTHER_INFORMATION: None,
            ATTR_DEVIATIONS: None,
    }
    if delay_in_minutes := self._state.get_delay_time():
        self._attr_extra_state_attributes[ATTR_DELAY_TIME] = (
            delay_in_minutes.total_seconds() / 60
        )
    if (time_at_location := self._state.advertised_time_at_location):
        self._attr_extra_state_attributes[ATTR_PLANNED_TIME] = (
            as_utc(time_at_location.astimezone(TIMEZONE)).isoformat()
        )
    ...
```